### PR TITLE
Event bus

### PR DESCRIPTION
**Pull request by VolodymyrM27**
@HNDaG

### CODE REVIEW

File: airguardian-core/src/main/kotlin/systems/ajax/motrechko/airguardian/controller/grpc/DroneBatteryApplicationGrpcService.kt [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+            successBuilder.newStateBuilder.setApplication(droneApplication)
+        }.build()
+
+    private fun buildSuccessInitialStateResponse(
+        droneApplicationList: List<DroneBatteryChargingApplication>
+    ): GetAllApplicationResponse {
+         var debug =  GetAllApplicationResponse.newBuilder().apply {
+            successBuilder.initialStateBuilder.addAllApplicationList(droneApplicationList)
+        }.build()
+        println(debug)
```
Comment (HNDaG) [inline airguardian-core/src/main/kotlin/systems/ajax/motrechko/airguardian/controller/grpc/DroneBatteryApplicationGrpcService.kt:53]: Remove this print

**Review by HNDaG — APPROVED** [review]
LGTM

File: internal-api/src/main/proto/systems/ajax/motrechko/airguardian/output/pubsub/application/drone_battery_charging_application_event.proto [hunk 2]
```diff
@@ -10,3 +10,7 @@ import "systems/ajax/motrechko/airguardian/commonresponse/application/drone_batt
 message DroneBatteryChargingApplicationEvent{
   commonresponse.application.drone_battery_charging_application.proto.DroneBatteryChargingApplication application = 1;
 }
+
+message DroneBatteryChargingApplicationEventList{
+  repeated  commonresponse.application.drone_battery_charging_application.proto.DroneBatteryChargingApplication applications = 1;
```
Comment (ajax-gnatiuk-s) [inline internal-api/src/main/proto/systems/ajax/motrechko/airguardian/output/pubsub/application/drone_battery_charging_application_event.proto:15]:
```suggestion
  repeated commonresponse.application.drone_battery_charging_application.proto.DroneBatteryChargingApplication applications = 1;
```

File: internal-api/src/main/proto/systems/ajax/motrechko/airguardian/drone_charge_application_service.proto [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package systems.ajax.motrechko.airguardian;
+
+import "systems/ajax/motrechko/airguardian/input/reqrepl/application/get_charge_application.proto";
+
+
+service ChargingDroneBatteryApplication {
+    rpc GetBatteryChargingApplication (input.reqrepl.application.get_charge_application.GetAllApplicationRequest)
```
Comment (ajax-gnatiuk-s) [inline internal-api/src/main/proto/systems/ajax/motrechko/airguardian/drone_charge_application_service.proto:7]:
```suggestion
    rpc getBatteryChargingApplication (input.reqrepl.application.get_charge_application.GetAllApplicationRequest)
```

File: airguardian-core/src/main/kotlin/systems/ajax/motrechko/airguardian/controller/nats/BatteryDroneChargingApplicationEventNatsController.kt [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+import io.nats.client.Dispatcher
+import reactor.core.publisher.Flux
+import systems.ajax.motrechko.airguardian.commonresponse.application.drone_battery_charging_application.proto.DroneBatteryChargingApplication
+
+interface BatteryDroneChargingApplicationEventNatsController<Event: GeneratedMessageV3> {
+
+    val parser: Parser<Event>
+
+    val dispatcher: Dispatcher
+    fun subscribeToEvent(eventType: String): Flux<Event>
```
Comment (ajax-gnatiuk-s) [inline airguardian-core/src/main/kotlin/systems/ajax/motrechko/airguardian/controller/nats/BatteryDroneChargingApplicationEventNatsController.kt:14]:
```suggestion
    val dispatcher: Dispatcher
    
    fun subscribeToEvent(eventType: String): Flux<Event>
```

File: internal-api/src/main/proto/systems/ajax/motrechko/airguardian/input/reqrepl/application/get_charge_application.proto [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+option java_multiple_files = true;
+
+import "systems/ajax/motrechko/airguardian/commonresponse/application/drone_battery_charging_application.proto";
+import "systems/ajax/motrechko/airguardian/output/pubsub/application/drone_battery_charging_application_event.proto";
+
+message GetAllApplicationRequest {
+
+}
+
+message GetAllApplicationResponse {
```
Comment (ajax-gnatiuk-s) [inline internal-api/src/main/proto/systems/ajax/motrechko/airguardian/input/reqrepl/application/get_charge_application.proto:14]:
```suggestion

message GetAllApplicationsResponse {
```

File: internal-api/src/main/proto/systems/ajax/motrechko/airguardian/input/reqrepl/application/get_charge_application.proto [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+      NewState newState = 2;
+    }
+
+    message InitialState {
+      output.pubsub.application.drone_battery_charging_application.proto.DroneBatteryChargingApplicationEventList applicationList_deprecated= 1 [deprecated = true];
+      repeated commonresponse.application.drone_battery_charging_application.proto.DroneBatteryChargingApplication applicationList = 2;
+    }
+
+    message NewState{
+      output.pubsub.application.drone_battery_charging_application.proto.DroneBatteryChargingApplicationEvent application_deprecated = 1 [deprecated = true];
```
Comment (ajax-gnatiuk-s) [inline internal-api/src/main/proto/systems/ajax/motrechko/airguardian/input/reqrepl/application/get_charge_application.proto:33]: Why you deprecated the new code?

File: internal-api/src/main/proto/systems/ajax/motrechko/airguardian/input/reqrepl/application/get_charge_application.proto [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+      InitialState initialState = 1;
+      NewState newState = 2;
+    }
+
+    message InitialState {
+      output.pubsub.application.drone_battery_charging_application.proto.DroneBatteryChargingApplicationEventList applicationList_deprecated= 1 [deprecated = true];
+      repeated commonresponse.application.drone_battery_charging_application.proto.DroneBatteryChargingApplication applicationList = 2;
+    }
+
+    message NewState{
```
Comment (ajax-gnatiuk-s) [inline internal-api/src/main/proto/systems/ajax/motrechko/airguardian/input/reqrepl/application/get_charge_application.proto:32]:
```suggestion
    message NewState {
```

File: internal-api/src/main/proto/systems/ajax/motrechko/airguardian/input/reqrepl/application/get_charge_application.proto [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+message GetAllApplicationResponse {
+  oneof response {
+    Success success = 1;
+    Failure failure = 2;
+  }
+
+  message Success {
+
+    oneof state{
```
Comment (ajax-gnatiuk-s) [inline internal-api/src/main/proto/systems/ajax/motrechko/airguardian/input/reqrepl/application/get_charge_application.proto:22]:
```suggestion
    oneof state {
```

File: internal-api/src/main/proto/systems/ajax/motrechko/airguardian/input/reqrepl/application/get_charge_application.proto [hunk 9]
```diff
@@ -0,0 +1,41 @@
+syntax = "proto3";
+
+package systems.ajax.motrechko.airguardian.input.reqrepl.application.get_charge_application;
+
+option java_multiple_files = true;
+
+import "systems/ajax/motrechko/airguardian/commonresponse/application/drone_battery_charging_application.proto";
+import "systems/ajax/motrechko/airguardian/output/pubsub/application/drone_battery_charging_application_event.proto";
+
+message GetAllApplicationRequest {
```
Comment (ajax-gnatiuk-s) [inline internal-api/src/main/proto/systems/ajax/motrechko/airguardian/input/reqrepl/application/get_charge_application.proto:10]:
```suggestion
message GetAllApplicationsRequest {
```

File: internal-api/src/main/proto/systems/ajax/motrechko/airguardian/input/reqrepl/application/get_charge_application.proto [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+  }
+
+  message Success {
+
+    oneof state{
+      InitialState initialState = 1;
+      NewState newState = 2;
+    }
+
+    message InitialState {
```
Comment (ajax-gnatiuk-s) [inline internal-api/src/main/proto/systems/ajax/motrechko/airguardian/input/reqrepl/application/get_charge_application.proto:27]:
```suggestion
    message ChargingApplicationUpdate {
```

File: airguardian-core/src/main/kotlin/systems/ajax/motrechko/airguardian/mapper/DroneBatteryApplicationMapper.kt [hunk 11]
```diff
@@ -1,11 +1,50 @@
...
+        BatteryApplicationStatus.COMPLETED -> ProtoBatteryApplicationStatus.COMPLETED
+    }
+}
+
+fun ProtoBatteryApplicationStatus.toEntity(): BatteryApplicationStatus{
+    return when (this) {
+        ProtoBatteryApplicationStatus.NEW -> BatteryApplicationStatus.NEW
+        ProtoBatteryApplicationStatus.IN_PROGRESS -> BatteryApplicationStatus.IN_PROGRESS
+        ProtoBatteryApplicationStatus.COMPLETED -> BatteryApplicationStatus.COMPLETED
+        else -> BatteryApplicationStatus.NEW
```
Comment (ajax-gnatiuk-s) [inline airguardian-core/src/main/kotlin/systems/ajax/motrechko/airguardian/mapper/DroneBatteryApplicationMapper.kt:48]: Actually, `else` should be mapped to default value, it should throw exception

File: airguardian-core/src/main/kotlin/systems/ajax/motrechko/airguardian/controller/nats/BatteryDroneChargingApplicationEventNatsController.kt [hunk 12]
```diff
@@ -0,0 +1,17 @@
+package systems.ajax.motrechko.airguardian.controller.nats
+
+import com.google.protobuf.GeneratedMessageV3
+import com.google.protobuf.Parser
+import io.nats.client.Dispatcher
+import reactor.core.publisher.Flux
+import systems.ajax.motrechko.airguardian.commonresponse.application.drone_battery_charging_application.proto.DroneBatteryChargingApplication
+
+interface BatteryDroneChargingApplicationEventNatsController<Event: GeneratedMessageV3> {
```
Comment (ajax-gnatiuk-s) [inline airguardian-core/src/main/kotlin/systems/ajax/motrechko/airguardian/controller/nats/BatteryDroneChargingApplicationEventNatsController.kt:9]: It's not controller, it's publisher

File: airguardian-core/src/main/kotlin/systems/ajax/motrechko/airguardian/controller/nats/BatteryDroneChargingApplicationEventNatsControllerImpl.kt [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+
+    override val parser: Parser<DroneBatteryChargingApplicationEvent> =
+        DroneBatteryChargingApplicationEvent.parser()
+
+    override val dispatcher: Dispatcher = connection.createDispatcher()
+    override fun subscribeToEvent(eventType: String): Flux<DroneBatteryChargingApplicationEvent> =
+        Flux.create { sink ->
+            dispatcher.apply {
+                subscribe(NatsSubject.BatteryDroneChargingApplication.GET_ALL)
+                { message ->
```
Comment (ajax-gnatiuk-s) [inline airguardian-core/src/main/kotlin/systems/ajax/motrechko/airguardian/controller/nats/BatteryDroneChargingApplicationEventNatsControllerImpl.kt:25]:
```suggestion
                subscribe(NatsSubject.BatteryDroneChargingApplication.GET_ALL) { message ->
```

File: internal-api/src/main/kotlin/systems/ajax/motrechko/airguardian/internalapi/NatsSubject.kt [hunk 14]
```diff
@@ -1,9 +1,16 @@
...
+
 object NatsSubject {
-    private const val REQUEST_PREFIX = "systems.ajax.motrechko.airguardian.input.request"
 
-    private const val EVENT_PREFIX = "systems.ajax.motrechko.airguardian.output.pubsub"
+    object BatteryDroneChargingApplication {
+        private const val BATTERY_DRONE_CHARGING_APPLICATION_PREFIX =
+            "$EVENT_PREFIX.battery_drone_charging_application"
+
+        const val GET_ALL = "$BATTERY_DRONE_CHARGING_APPLICATION_PREFIX.get_all"
```
Comment (ajax-gnatiuk-s) [inline internal-api/src/main/kotlin/systems/ajax/motrechko/airguardian/internalapi/NatsSubject.kt:12]: "Get all" - is something to do.
But we need "something has happened".

File: internal-api/src/main/proto/systems/ajax/motrechko/airguardian/output/pubsub/application/drone_battery_charging_application_event.proto [hunk 15]
```diff
@@ -10,3 +10,7 @@ import "systems/ajax/motrechko/airguardian/commonresponse/application/drone_batt
 message DroneBatteryChargingApplicationEvent{
   commonresponse.application.drone_battery_charging_application.proto.DroneBatteryChargingApplication application = 1;
 }
+
+message DroneBatteryChargingApplicationEventList{
```
Comment (ajax-gnatiuk-s) [inline internal-api/src/main/proto/systems/ajax/motrechko/airguardian/output/pubsub/application/drone_battery_charging_application_event.proto:14]:
```suggestion
message DroneBatteryChargingApplicationEventList {
```

File: internal-api/src/main/proto/systems/ajax/motrechko/airguardian/drone_charge_application_service.proto [hunk 16]
```diff
@@ -0,0 +1,9 @@
+package systems.ajax.motrechko.airguardian;
+
+import "systems/ajax/motrechko/airguardian/input/reqrepl/application/get_charge_application.proto";
+
+
+service ChargingDroneBatteryApplication {
+    rpc getBatteryChargingApplication (input.reqrepl.application.get_charge_application.GetAllApplicationsRequest)
```
Comment (ajax-gnatiuk-s) [inline internal-api/src/main/proto/systems/ajax/motrechko/airguardian/drone_charge_application_service.proto:7]:
```suggestion
    rpc getBatteryChargingApplication(input.reqrepl.application.get_charge_application.GetAllApplicationsRequest)
```

File: airguardian-core/src/main/kotlin/systems/ajax/motrechko/airguardian/controller/nats/BatteryDroneChargingApplicationEventNatsPublisher.kt [hunk 17]
```diff
@@ -0,0 +1,17 @@
...
+import io.nats.client.Dispatcher
+import reactor.core.publisher.Flux
+import systems.ajax.motrechko.airguardian.commonresponse.application.drone_battery_charging_application.proto.DroneBatteryChargingApplication
+
+interface BatteryDroneChargingApplicationEventNatsPublisher<Event: GeneratedMessageV3> {
+
+    val parser: Parser<Event>
+
+    val dispatcher: Dispatcher
+    fun subscribeToEvent(eventType: String): Flux<Event>
```
Comment (ajax-gnatiuk-s) [inline airguardian-core/src/main/kotlin/systems/ajax/motrechko/airguardian/controller/nats/BatteryDroneChargingApplicationEventNatsPublisher.kt:14]:
```suggestion
    val dispatcher: Dispatcher
    
    fun subscribeToEvent(eventType: String): Flux<Event>
```

File: airguardian-core/src/main/kotlin/systems/ajax/motrechko/airguardian/controller/grpc/DroneBatteryApplicationGrpcService.kt [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+import systems.ajax.motrechko.airguardian.output.pubsub.application.drone_battery_charging_application.proto.DroneBatteryChargingApplicationEvent
+import systems.ajax.motrechko.airguardian.service.DroneBatteryService
+
+@GrpcService
+class DroneBatteryApplicationGrpcService(
+    private val droneBatteryService: DroneBatteryService,
+    private val batteryDroneChargingApplicationEventNatsController:
+    BatteryDroneChargingApplicationEventNatsPublisher<DroneBatteryChargingApplicationEvent>,
+) : ReactorChargingDroneBatteryApplicationGrpc.ChargingDroneBatteryApplicationImplBase() {
+    override fun getBatteryChargingApplication(request: Mono<GetAllApplicationsRequest>)
```
Comment (ajax-gnatiuk-s) [inline airguardian-core/src/main/kotlin/systems/ajax/motrechko/airguardian/controller/grpc/DroneBatteryApplicationGrpcService.kt:22]:
```suggestion
) : ReactorChargingDroneBatteryApplicationGrpc.ChargingDroneBatteryApplicationImplBase() {

    override fun getBatteryChargingApplication(request: Mono<GetAllApplicationsRequest>)
```

File: airguardian-core/src/main/kotlin/systems/ajax/motrechko/airguardian/mapper/DroneBatteryApplicationMapper.kt [hunk 19]
```diff
@@ -1,11 +1,50 @@
...
+        BatteryApplicationStatus.COMPLETED -> ProtoBatteryApplicationStatus.COMPLETED
+    }
+}
+
+fun ProtoBatteryApplicationStatus.toEntity(): BatteryApplicationStatus{
+    return when (this) {
+        ProtoBatteryApplicationStatus.NEW -> BatteryApplicationStatus.NEW
+        ProtoBatteryApplicationStatus.IN_PROGRESS -> BatteryApplicationStatus.IN_PROGRESS
+        ProtoBatteryApplicationStatus.COMPLETED -> BatteryApplicationStatus.COMPLETED
+        else -> throw IllegalArgumentException("Unknown BatteryApplicationStatus")
```
Comment (ajax-gnatiuk-s) [inline airguardian-core/src/main/kotlin/systems/ajax/motrechko/airguardian/mapper/DroneBatteryApplicationMapper.kt:48]: **NIT**
```suggestion
        else -> error("Unknown BatteryApplicationStatus: $this")
```

File: internal-api/src/main/kotlin/systems/ajax/motrechko/airguardian/internalapi/NatsSubject.kt [hunk 20]
```diff
@@ -1,9 +1,16 @@
...
+
 object NatsSubject {
-    private const val REQUEST_PREFIX = "systems.ajax.motrechko.airguardian.input.request"
 
-    private const val EVENT_PREFIX = "systems.ajax.motrechko.airguardian.output.pubsub"
+    object BatteryDroneChargingApplication {
+        private const val BATTERY_DRONE_CHARGING_APPLICATION_PREFIX =
+            "$EVENT_PREFIX.battery_drone_charging_application"
+
+        const val PUBLISH_NEW_APPLICATION = "$BATTERY_DRONE_CHARGING_APPLICATION_PREFIX.get_all"
```
Comment (ajax-gnatiuk-s) [inline internal-api/src/main/kotlin/systems/ajax/motrechko/airguardian/internalapi/NatsSubject.kt:12]: let's discuss offline

**Review by ajax-gnatiuk-s — APPROVED** [review]

**Pull request merged**